core/src/main/java/org/apache/calcite/plan/volcano/RelSubset.java

```diff
@@ -319,9 +321,67 @@ RelNode buildCheapestPlan(VolcanoPlanner planner) {
    */
   void propagateCostImprovements(VolcanoPlanner planner, RelMetadataQuery mq,
       RelNode rel, Set<RelSubset> activeSet) {
+//    for (RelSubset subset : set.subsets) {
+//      if (rel.getTraitSet().satisfies(subset.traitSet)) {
+//        subset.propagateCostImprovements0(planner, mq, rel, activeSet);
+//      }
+//    }
+    hierarchicalPropagateImprovement(planner, mq, this, rel, activeSet);
+  }
+
+  void propagateCostImprovements0(VolcanoPlanner planner, RelMetadataQuery mq,
+      RelNode rel, Set<RelSubset> activeSet,
+      Stack<Pair<RelNode, RelSubset>> propagateStack) {
+    ++timestamp;
+
+    if (!activeSet.add(this)) {
+      // This subset is already in the chain being propagated to. This
+      // means that the graph is cyclic, and therefore the cost of this
+      // relational expression - not this subset - must be infinite.
+      LOGGER.trace("cyclic: {}", this);
+      return;
+    }
+    try {
+      final RelOptCost cost = planner.getCost(rel, mq);
+      if (cost.isLt(bestCost)) {
+        LOGGER.trace("Subset cost improved: subset [{}] cost was {} now {}",
+            this, bestCost, cost);
+
+        bestCost = cost;
+        best = rel;
+
+        // Lower cost means lower importance. Other nodes will change
+        // too, but we'll get to them later.
+        planner.ruleQueue.recompute(this);
+        for (RelNode parent : getParents()) {
+          final RelSubset parentSubset = planner.getSubset(parent);
+          parentSubset.push(parent, propagateStack);
+        }
+        planner.checkForSatisfiedConverters(set, rel);
+      }
+    } finally {
+      activeSet.remove(this);
+    }
+  }
+
+  void hierarchicalPropagateImprovement(VolcanoPlanner planner, RelMetadataQuery mq,
+      RelSubset subset, RelNode rel, Set<RelSubset> activeSet) {
+    Stack<Pair<RelNode, RelSubset>> propagateStack = new Stack<>();
```

Comment:
Stack class is very legacy, it extends Vector, maybe we can use some better util class

Reply:
ok, It only need List for FIFO instead of Stack